### PR TITLE
added py files for kernel ridge and assessor

### LINKED ISSUE #54: Implement ML model

### LINKED ISSUE #74: Create RFF function to speed up implementation of Gaussian kernel

### PR DESCRIPTION
**Pull request by sanjeevdasgupta**

### CODE REVIEW

**Comment by ajrichter7** [comment]
I noticed a few bugs in the codes, but nothing so bad should I merge and change them? @sanjeevdasgupta

I believe in the assessment code I wrote I need to change the output, I have it returning nparray but I designated that the functions should return a value. But overall I think it is okay, the Guassian kernel seems correct but for fit i believe we are missing a return statement to exit the function. Do we have all the necessary class functions for the Kernel Ridge? Did I miss any for functionality?

File: app/assessor.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+        Args: 
+            y_test (np.ndarray): true values for the outcome 
+            y_pred (np.ndarray): predicted values for the outcome given our model 
+
+        Returns: 
+            MSE (float): the average squared difference between the actual and predicted values 
+        '''
+        score = np.square(np.subtract(y_test, y_pred)).mean()
+        self.cv_scores.append(score)
```
Comment (ajrichter7) [inline app/assessor.py]: Not sure if we need to do the cross validation scores like this. Either we do this or we return just the float. Could add that as a different element for the model initializer class, what do you think is best?

File: app/assessor.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            y_pred (np.ndarray): predicted values for the outcome given our model 
+
+        Returns: 
+            accuracy (float): r^2 score for a given model which is a number a <= 1 (with 1 being best)
+        '''
+        mean_y_test = np.mean(y_test)
+        TSS = np.sum((y_test - mean_y_test) ** 2)
+        RSS = np.sum((y_test - y_pred) ** 2)
+        score = 1 - (RSS / TSS) 
+        self.cv_scores.append(score)
```
Comment (ajrichter7) [inline app/assessor.py]: Similar comment as for MSE but basically this returns numpy array and not a float

File: app/assessor.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+
+class Assessment:
+    def __init__(self):
+        self.cv_scores = []
+    
+    def r2_score(self, y_test:np.ndarray, y_pred:np.ndarray) -> float:
```
Comment (ajrichter7) [inline app/assessor.py:5]: Return value

File: app/assessor.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            accuracy (float): r^2 score for a given model which is a number a <= 1 (with 1 being best)
+        '''
+        mean_y_test = np.mean(y_test)
+        TSS = np.sum((y_test - mean_y_test) ** 2)
+        RSS = np.sum((y_test - y_pred) ** 2)
+        score = 1 - (RSS / TSS) 
+        self.cv_scores.append(score)
+        return np.array(self.cv_scores)
+
+    def mean_squared_error(self, y_test:np.ndarray, y_pred:np.ndarray) -> float:
```
Comment (ajrichter7) [inline app/assessor.py:25]: return value

File: app/kernel_ridge.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            y_train (np.ndarray): Training target values.
+
+        Returns: 
+            not so much as setting the dual coef that we will use to predict 
+        """
+        
+        K_train = gaussian_kernel(X_train, X_train, sigma2=self.sigma2)
+        self.coef_ = np.linalg.solve(K_train + self.lamb * np.eye(len(X_train)), y_train)
+        self.X_train_ = X_train
+
```
Comment (ajrichter7) [inline app/kernel_ridge.py:71]: return statement needed

File: app/kernel_ridge.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import numpy as np
+
```
Comment (ajrichter7) [inline app/kernel_ridge.py:3]: Do we need any more functions for kernel ridge?

**Review by ajrichter7 — CHANGES_REQUESTED** [review]
I reviewed the code and noticed a few small issues as I mentioned in an earlier comment. Overarching would just be is there more to be done for the kernel class or gaussian kernel coding wise?

Maybe for r2_score / MSE we could add another function to the assessment class that adds the score to a cv list or np array? What do you think is best?

File: app/assessor.py (hunk 1 above)
Comment (sanjeevdasgupta) [reply app/assessor.py]: Agree - I think the float is enough. I also verified this via the tests Santiago ran. Have updated the code accordingly.

File: app/assessor.py (hunk 2 above)
Comment (sanjeevdasgupta) [reply app/assessor.py]: Agree - same as above. Have updated.

File: app/assessor.py (hunk 3 above)
Comment (sanjeevdasgupta) [reply app/assessor.py:5]: This is the same point as earlier right? Or was there a different thing that you are proposing we return?

File: app/assessor.py (hunk 4 above)
Comment (sanjeevdasgupta) [reply app/assessor.py:25]: Same question as above - I am slightly unclear how this is different from the previous comment

File: app/kernel_ridge.py (hunk 5 above)
Comment (sanjeevdasgupta) [reply app/kernel_ridge.py:71]: I checked this with Santiago's tests and it seems to work without the return statement as well. Nevertheless, I've added a return statement that returns the coefficients

File: app/kernel_ridge.py (hunk 6 above)
Comment (sanjeevdasgupta) [reply app/kernel_ridge.py:3]: In general from Santiago's tests I think we should be okay. However, I think for the final predicted values, we should also show uncertainty intervals to help with interpretation. I looked into some options and we could do either bootstrapping or a Bayesian prediction interval. Bootstrapping would require us to re-run the model several times, however, which I think won't be feasible for a web app. So I've added a function to calculate prediction intervals using the Bayesian approach. I've also made this more efficient by calculating the matrix inverse we require for the approach during model training so that they it doesn't have to be inverted during the call from the web app.

**Comment by sanjeevdasgupta** [comment]
Have updated the code based on feedback and pushed changes again. I have also added the random Fourier features approximation to see if that is faster than KRR like we discussed. @ajrichter7 if everything looks okay to you, please go ahead and approve the changes and merge.

File: app/assessor.py (hunk 4 above)
Comment (ajrichter7) [reply app/assessor.py:25]: It is not different. I just tried to comment on every instance. I was just making sure you saw. :)

File: app/kernel_ridge.py (hunk 5 above)
Comment (ajrichter7) [reply app/kernel_ridge.py:71]: Should we return the coef or self? I looked at the sklearn source code and they return the object. I don't think that we need to return the coef matrix. My statement was more so a python statement about how methods should have a return statement, but I also did just [read here ](https://realpython.com/python-return-statement/) about implicit return statements. It was a comment about readability and best practice.

File: app/assessor.py (hunk 3 above)
Comment (ajrichter7) [reply app/assessor.py:5]: same as earlier

File: app/assessor.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+
+class Assessment:
+    def __init__(self):
```
Comment (ajrichter7) [inline app/assessor.py]: We can delete this

Comment (ajrichter7) [reply app/assessor.py]: We are not initializing any values for this class. [Stack overflow says not needed if we not initializing anything anymore. ](https://stackoverflow.com/questions/6854080/is-it-necessary-to-include-init-as-the-first-function-every-time-in-a-class)

File: app/kernel_ridge.py (hunk 5 above)
Comment (ajrichter7) [reply app/kernel_ridge.py:71]: After some looking, I think we could do the implicit return if you say that works in Santiago, I just was saying to return self. But reading some blogs, I think that maybe it is more pythonic to just not include it since we set an attribute. I don't think we need to return the coef explicitly since it is set and stored in the class. But also, if you compared with Santiago's point and it works with the implicit return maybe we can close this and return to it later. Close as in no longer explicitly return self.coef. Sorry for raising this earlier, I had to look into best coding practices for python classes.

File: app/random_fourier_features.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+import numpy as np
+
+def rff_features(
+    x: np.ndarray,
+    R: int,
+    sigma: float = 1.0,
+    seed: int | None = None,
+) -> np.ndarray:
+    r"""Compute Random Fourier Features approximating the Gaussian kernel.
```
Comment (ajrichter7) [inline app/random_fourier_features.py]: The formatting here with :math:` and r""" is a little inconsistent with other functions/classes we have defined. Sorry it is a small note

Comment (ajrichter7) [reply app/random_fourier_features.py]: I mean within the entire explanation text

File: app/random_fourier_features.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            y_train (np.ndarray): Training target values.
+            R (int): Number of random Fourier features to run
+
+        Returns: 
+            Sets and returns the coefficients
+        """
+
+        Z_train = rff_features(X_train, R, sigma = self.sigma, seed = seed)
+        self.coef_ = np.linalg.solve(Z_train.T @ Z_train + self.lamb * np.eye(R), Z_train.T @ y_train)
+        return self.coef_
```
Comment (ajrichter7) [inline app/random_fourier_features.py]: Sorrryy see my note above, but I self reflected haha. Setters need not have return explicit of the the thing. I also don't think wrong, its a lot of "what is best practice" and I was wrong initially. I think the best practice for python is to not return for setters

Comment (ajrichter7) [reply app/random_fourier_features.py]: Here the [source code for sklearn](https://github.com/scikit-learn/scikit-learn/blob/fe2edb3cdbd75ae4e662fda67dcb19277258792b/sklearn/gaussian_process/_gpr.py) does the return self. I got a bit mixed up because it is like a long function that is in essence just a setter.

File: app/random_fourier_features.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            X_train (np.ndarray): Training feature data. 
+            y_train (np.ndarray): Training target values.
+            R (int): Number of random Fourier features to run
+
+        Returns: 
+            Sets and returns the coefficients
+        """
+
+        Z_train = rff_features(X_train, R, sigma = self.sigma, seed = seed)
+        self.coef_ = np.linalg.solve(Z_train.T @ Z_train + self.lamb * np.eye(R), Z_train.T @ y_train)
```
Comment (ajrichter7) [inline app/random_fourier_features.py:65]: I spent a while thinking about this and convinced myself of this math after [reading this blog](https://gregorygundersen.com/blog/2019/12/23/random-fourier-features/). Equation 24. I think this implementation makes sense. No work just a commenting.

File: app/random_fourier_features.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+    so that :math:`Z Z^\top \approx K` for the Gaussian kernel with
+    bandwidth :math:`\sigma`.
+    """
+    rng = np.random.RandomState(seed) if seed is not None else np.random
+    x = np.asarray(x, dtype=float).ravel()
+    W = rng.randn(R)  # (R,)
+    B = rng.uniform(0, 2 * np.pi, size=R)  # (R,)
+    Z = np.sqrt(2.0 / R) * np.cos(sigma * np.outer(x, W) + B[None, :])
+    return Z
```
Comment (ajrichter7) [inline app/random_fourier_features.py:26]: This is what drew has in his slides I think I am only getting a little confused on when things are transposed but I think this implementation works, based on the cited blog below and Lecture 05 slide 32. We can test it anyways when we get the models working.

File: app/random_fourier_features.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    :math:`U(0, 2\pi)`, then returns the (n, R) feature matrix
+
+    .. math::
+
+        Z = \sqrt{2/R}\,\cos(\sigma\,W\,X^\top + B)
+
+    so that :math:`Z Z^\top \approx K` for the Gaussian kernel with
+    bandwidth :math:`\sigma`.
+    """
+    rng = np.random.RandomState(seed) if seed is not None else np.random
```
Comment (ajrichter7) [inline app/random_fourier_features.py]: I see and understand this I originally commented on the initialization of the arguments because it is inconsistent. I don't think it is wrong by any means but just normally we would set it as 42 if no seed is passed, but this time you say if none then random. Just pointing out the inconsistency nothing wrong.

File: app/kernel_ridge.py (hunk 6 above)
Comment (ajrichter7) [reply app/kernel_ridge.py:3]: coolio thanks

**Review by ajrichter7 — CHANGES_REQUESTED** [review]
Hey, I left a few comments. Sorry for being verbose sometimes. I think maybe if you just change the return statement back and some minor comment formatting. (I am sorry about getting confused on the return statements and proper coding practice). Sorry it took me a bit of time, I wanted to read it through and think on it all. But the only two changes are super small, but I will approve right away. Should take like no time.

**Comment by sanjeevdasgupta** [comment]
Just respond with one comment here so I don't bombard you with multiple email notifications again :)

Thanks a lot for the second round of feedback! I agree with all of your suggestions so I've fixed the following:

1. Removed the return for the self coefficient setter across both kernel and Fourier
2. Remove the __init__ for assessor
3. Cleaned up the formatting for the description of the Fourier function so that it now matches other description formatting (basically removing the r and the math formatting, both of which I picked up from Drew's code)
4. Made the seed set up consistent so that it now always defaults to 42 instead of random

Let me know if you still spot anything. Otherwise, hopefully we should be good to merge now!
